src/signature/spki.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright 2015 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR
+// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+use der;
+use signature;
+use untrusted;
+
+/// An error that occurs during certificate validation or name validation.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum VerifyWithSPKIError {
+    /// The encoding of some ASN.1 DER-encoded item is invalid.
+    BadDER,
+
+    /// The signature is invalid for the given public key.
+    InvalidSignatureForPublicKey,
+
+    /// The SignatureAlgorithm does not match the algorithm of the SPKI.
+    /// A mismatch could be because of the algorithm (RSA vs DSA, etc) or the
+    /// parameters (ECDSA_p256 vs ECDSA_384, etc).
+    UnsupportedSignatureAlgorithmForPublicKey,
+}
+
+/// Verify the signature `signature` of message `msg` with the public key
+/// `spki_public_key` using the algorithm `alg`.
+/// `signature::verify` is suitable when the public key bytes you have are the
+/// DER ASN.1 of the key. This method is suitable when the public key is in
+/// SubjectPublicKeyInfo form (https://tools.ietf.org/html/rfc5280#section-4.1),
+/// which is a wrapper containing both the key and an identifier.
+///
+/// A common situation where this encoding is encountered is when using public keys
+/// exported by OpenSSL. If you export an RSA or ECDSA public key from a keypair
+/// with `-pubout` and friends, you will get a DER of an SPKI wrapper containing
+/// a public key. You could extract the bitstring of the key itself and use
+/// `signature:;verify`, but that is often inconvient.
+pub fn verify(signature_alg: &Algorithm,
```

Comment:
I think we're on the same page, but just to be sure I'll elaborate.

Tracing the data flow up from `verify_signed_data`:
```
EndEntityCert::verify_is_valid_tls_server_cert // top of call stack, takes a &[TrustAnchor] 
verify_cert::build_chain // takes a &[TrustAnchor] 
verify_cert::check_signatures // called on TrustAnchor.spki
signed_data::verify_signed_data 
signed_data::verify_signature // function now in ring
```

And `TrustAnchor` is created in `trust_anchor_util::parse_cert_v1` with an `spki` of:

```
let spki =
                    try!(der::expect_tag_and_get_value(tbs,
                                                       der::Tag::Sequence));
```

or via `trust_anchor_from_cert` in `parse_cert_internal` with identical code.

webpki's representation of spki as just the contents of the sequence is sensible because it saves a few mostly-meaningless bytes, especially when all the code that uses it is internal. 

I think the choices are as follows:

1. Leave webpki's representation as it is, and have *ring*'s function continue to expect that as input. For users that haven't already stripped off the outer sequence (as when they're loading a key from a file), document the verification function to explain that `expect_tag_and_get_value` should be called first on the DER.
1. Switch webpki to keep the complete SPKI DER internally, and switch *ring*'s function accordingly. Now both webpki and average joes working with a public key have no intervening steps to process the key data from the point of loading the bytes. This is nice cleanliness-wise, but costs a few bytes of memory to hang onto the key, and a few cycles for every verification to strip off the sequence. 

Ironically given how I started this feature request, I'm slightly inclined towards the former choice given the low-level nature and performance sensitivity of this library, but the latter is also pleasingly tidy. Either seems quite defensible, though. I *think* you're suggesting going with the second option; is that right?